### PR TITLE
Change BlockReader API

### PR DESCRIPTION
**Pull request by hmacdope**
This is an updated version of the CythonBlockReader API, incorporating some changes that @richardjgowers  suggested.

- Primary change is that you now need to provide a numpy array to read block data into rather than it being tied to the block instance.

Still not quite complete as I need to improve the array copying and add some memoryviews.

### CODE REVIEW

**Comment by hmacdope** [comment]
ready for review :)

**Comment by orbeckst** [comment]
@hmacdope impressive progress. Sorry, my review was hastily written and does not convey how happy I am that this appears to be at an usable state.

**Comment by hmacdope** [comment]
Thanks for the feedback @orbeckst, I'll work through bit by bit over coming days.

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:1327]: Correct, perhaps @tylerjereddy would know If this can be done better?

Other options are:
* tying the C array to a `MemoryWrapper` instance and then using PyArray_SetBaseObject, meaning that we do the memory management for the caller (IMO bad idea).
* Iterate over the C array and put it in the NumPy array (FLOP inefficent)
* ??

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:1287]: My bad, hangover from this being nogil. Ill fix.

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:1270]: Yes as we have GIL here. :)

File: pytng/pytng.pyx
Comment (tylerjereddy) [inline pytng/pytng.pyx:1254]: Are any of these typed memoryviews going to hold C-contiguous data? You've probably thought of that already, but can maybe use the `::1` notation to squeeze out a little more perf under the hood.

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:59]: gone now as misleading and will only return last frame.

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst]: I think `traj.tng` should be clear enough.

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx]: There is a fair bit of junk hanging around to support the old API. Should I make a separate PR to strip it all out or clean up here?

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx]: I think new structure resolves this, but would be grateful if you could take a quick peek. :)

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:984]: no problem, I think that is a sensible pattern for access?

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:79]: we are now both reading in place and returning the value ie:

`positions = ts.get_positions(positions)`

and 

`ts.get_positions(positions)`

both work.

Is this how we want the final product to look, or would we prefer just a single method to be canonical?

Comment (hmacdope) [reply docs/index.rst:79]: or do we want to rename the getter as well :)

Comment (orbeckst) [reply docs/index.rst:79]: I like the returning values in addition to the in-place. 

I can live with `get_*()` if you feel that this is the best name. After all, this is more your code than mine ;-).

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:32]: Additionally

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:26]: e.g.,

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:40]: link 
```reST
:class:`TNGFileIterator`
```

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:71]: Just a question: is there a way to iterate over all steps that contain any data, e.g., if block_strides for positions are 100 and 1000 for velocities then we would automatically do stride 100 and for 9 out of 10 steps, we just don't get velocity data.

More weirdly, when positions are saved every 5 steps and velocities every 11 steps then we would get only a few steps that have both (at step = n * 5 * 11, n = 0, 1, ,2, ...).

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:100]: Are there any specific exceptions that the user needs to be aware of?

Are exceptions documented in the classes (sorry, I am short on time reviewing this)

File: pytng/pytng.pyx
Comment (orbeckst) [inline pytng/pytng.pyx:897]: I think you're mixing doc styles here. I the `:rtype` line is superfluous.

File: pytng/pytng.pyx
Comment (orbeckst) [inline pytng/pytng.pyx:1387]: Is this supposed to be 
```reST
:data:`TNG_CRITICAL`
```
Not clear from the docs how the user can check for the status code.

File: pytng/pytng.pyx
Comment (orbeckst) [inline pytng/pytng.pyx:597]: remove? is the frame tuple still used?

File: pytng/pytng.pyx
Comment (orbeckst) [inline pytng/pytng.pyx:1456]: globals... really?

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:93]: What other data can be read? Explain velocity and forces.

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:77]: Include box vectors in example.

File: docs/index.rst
Comment (orbeckst) [reply docs/index.rst:93]: What happens if you go to a integrator step that does not contain data? How to deal with this case?

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:88]: Why numpy here? You're not using it (and presumably assume that `positions` is already allocated)

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:71]: The easiest solution I thought of was to stride over the greatest common divisor of the strides.
This is already computed as `TngFileIterator._gcd`, but I haven't made use of it yet.

Does that sound reasonable?

Comment (orbeckst) [reply docs/index.rst:71]: The GCD(5, 11) = 1 so what does this mean in terms of iterating? We get a lot of empty steps?

Comment (orbeckst) [reply docs/index.rst:71]: For the likely more common case of 10,100 the GCD sounds pretty sane, i.e., when GCD(m, n) == m or n. In the case where we get empty steps then the calling code will need to decide what to do (such as not returning a step) and that will be rather costly.

I'd much prefer if TNGFileIterator never returned empty frames. It tags frames so user code will just have to work with the frame metadata.

From MDAnalysis' perspective I also want to know what `len(tng)` should mean; I think the only sensible interpretation is "total number of non-empty frames" (like TRR).

Any opinions @richardjgowers @jbarnoud @kain88-de ?

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:597]: I will remove the old API once we get this one in I think?

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:76]: Working with steps is closer to the way the format itself works. My suggestion is that working in the original units of the format is better, we can then provide the user the tools they need to iterate with the minimum of wasted steps (obviously with big improvements in docs required).

I am happy to change it up though! No dramas.

I see where your coming from in regards to the length, would this mean the total number of frames with any block attached would become the length? This would still result in blank reads although a lot fewer. The other option is to have a separate length in each block "dimension".

Open to suggestion!

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:119]: see discussion above

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:87]: Hey @orbeckst I have tried to clear this up a bit and add more info in the docs.

For blank reads we currently fill the array with np.nan. Not the most elegant solution but gets the point across.  I have added an example that demonstrates this to the docs.

File: docs/index.rst
Comment (tylerjereddy) [inline docs/index.rst:38]: typo "effectivley"

File: docs/index.rst
Comment (tylerjereddy) [inline docs/index.rst:43]: In simpler terms this means that PyTNG does not support trajectories where the number of particles varies between frames?

File: docs/index.rst
Comment (tylerjereddy) [inline docs/index.rst:45]: One of the selling points of the TNG format when it was originally announced was that we wouldn't have to worry about keeping track of separate topology and trajectory files. Is the plan for MDAnalysis to eventually support a single TNG file as the source of both topology and trajectory?

File: tests/test_tng.py
Comment (tylerjereddy) [inline tests/test_tng.py:134]: `np.empty` vs. the low-level instantiator?

File: tests/test_tng.py
Comment (tylerjereddy) [inline tests/test_tng.py:151]: ok, so it looks like you can use the high-level NumPy array instantiators just fine then

File: tests/test_tng.py
Comment (tylerjereddy) [inline tests/test_tng.py:159]: is it worth parametrizing the indices tested to include a `-1` for reverse index support to access the last element?

File: tests/test_tng.py
Comment (tylerjereddy) [inline tests/test_tng.py:169]: will `-1`, `-9` indices work above as well?

File: tests/test_tng.py
Comment (tylerjereddy) [inline tests/test_tng.py:357]: Is there a test for handling "invalid" `dtype` for the array used to store the data? If you feed in `np.complex128` or a unicode type or something?

File: pytng/pytng.pyx
Comment (tylerjereddy) [inline pytng/pytng.pyx:1216]: To clarify, when it says the shape should be `(n_atoms, n_values_per_frame)` is that different from `(n_atoms, ndim)`? I think when I was reading through the examples at the start I was under the impression this was the number of dims? Maybe it is meant to be the same thing.

File: pytng/pytng.pyx
Comment (tylerjereddy) [inline pytng/pytng.pyx:1245]: As I was looking through the tests I was wondering what the supported dtypes for these input arrays might be, and if an exception would be raised for i.e., non-numeric dtypes, etc.

Looking farther down below--is this all routed through the dtype check in `get_blockid()` then?

**Review by tylerjereddy — COMMENTED** [review]
I added a few minor comments/questions since my review was requested again, but none of the comments are intended to be blocking for the merge; just suggestions or things I maybe didn't get.

I see that codecov is adding some tags for Cython lines that are not covered by tests yet; on the other hand, many foundational libraries don't even check the coverage of their Cython lines so not sure we have to be super strict on that, at least initially because this is a lot of work.

File: pytng/pytng.pyx
Comment (tylerjereddy) [reply pytng/pytng.pyx]: It certainly looks cleaner than the hard copy from before; I'll mark as resolved.

File: README.rst
Comment (hmacdope) [reply README.rst:26]: Ok noted!

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:45]: For sure thats the way forward. Issue raised to this effect in #44.

The file contains the whole gmx mtop, so would probably be able to leverage or adapt some existing MDA code.

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:1245]: I have tried to clarify available dtypes in the docs.

Yes you are right it all passes through that dtype check, and an exception is raised if incorrect.

**Comment by hmacdope** [comment]
Thanks for the review @tylerjereddy. Great points I will address ASAP.

File: tests/test_tng.py
Comment (hmacdope) [reply tests/test_tng.py:176]: Negative indices not currently supported. I should add them in.

Is there a canonical way to do that with an Iterator or should I implement my own version?

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:1216]: I think meant to be the same thing.

I think name `n_values_per_frame` (used in the library itself) is meant to provide the idea that for non-particle dependent blocks e.g box vectors we should expect that many values.

Open to changing if you think it would be clearer :)

File: tests/test_tng.py
Comment (hmacdope) [reply tests/test_tng.py:197]: now they will.:)

File: tests/test_tng.py
Comment (hmacdope) [reply tests/test_tng.py:176]: Fixed for proper wraparound on the iterator.

File: pytng/pytng.pyx
Comment (orbeckst) [reply pytng/pytng.pyx:597]: Do you mean it should be removed as part of this PR or that once this PR is in you can work on other things to eventually remove it?

File: docs/index.rst
Comment (orbeckst) [reply docs/index.rst:95]: Is processing np.nan expensive? 

In any case, I don't have a better idea right away (I guess we can't temporarily set the whole array to NULL or None?) Is there a way for calling code to quickly learn what data are available – is it just `np.all(np.isnan())`, which would be expensive? Could you add an example that shows how calling code can find out efficiently which data it can pull out?

Comment (orbeckst) [reply docs/index.rst:95]: The example above is for the simple case when the strides coincide. That's good to introduce the basics.

Now add an example where the strides differ. If I don't know how to do this I have no idea how write a Reader for e.g. MDA.

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:108]: Stay consistent... I was confused "What, this one returns a value and the other does in place?" until I read the comment. Just make both code examples "in place" and then show in the comments the return value form.

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:149]: This is the example that I asked for in an earlier comment (when I hadn't seen this one). 

The question remains how expensive `np.all(np.isnan(positions))` is. Don't the TNG iterator "know" that it filled the array and couldn't it set a flag that could be queried by calling code? Then we wouldn't even have to fill the array, calling code would just have to check the flag.

File: tests/test_tng.py
Comment (orbeckst) [reply tests/test_tng.py:197]: Add documentation, e.g. saying that standard Python slices are supported.

Are `slice()` objects also supported?

**Comment by orbeckst** [comment]
@hmacdope I added a few more hastily written comments. Coming along nicely, I don't want to overly hold up proceedings so please feel free to push for moving forward and/or suggesting to move forward with as is and raise other issues as necessary.

File: pytng/pytng.pyx
Comment (hmacdope) [reply pytng/pytng.pyx:597]: Sorry should have been clearer.

I'll remove the old API in a separate PR once this PR is merged. There is a lot of stuff that can be cleaned up but we have to remove it all at once so it can all still compile.

Comment (orbeckst) [reply pytng/pytng.pyx:597]: Thank you!

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:149]: I can add a query-able flag on the integrator step, or alternatively we can possible return a `NamedTuple` with the read status contained as one of the elements.

 I'll play with a few options :)

Comment (hmacdope) [reply docs/index.rst:163]: There is now a query-able flag for the success of the last read in `CurrentIntegratorStep.read_success`. Open to other suggestions :)

I think we should still fill the target array with `np.nan` for blank reads, otherwise the current contents of the target array will be the last read step which is potentially misleading when iterating.

I'll do some timings to see how costly filling with `np.nan` is.

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:202]: Explain what "successful" means. In particular, can it fail because (1) EOF, (2) file corruption, (3) no data stored for the step? Are these all equally unsuccessful?

Comment on reST: use double-backticks to format as code. Single backtick indicates a role in reST/sphinx (such as 
```reST
This indicates a class :class:`Universe` with the ":class:" role.
If I write `Universe` then it assumes the default role (whatever that is).
To explicitly mark up code use double back-ticks, e.g., ``u = mda.Universe(...)``, or ``(True != False) == True`` is ``True``.
```

and it just uses the default role

File: docs/index.rst
Comment (orbeckst) [reply docs/index.rst:163]: I'd be curious about the penalty for filling the array. It seems wasteful to do.

Comment (orbeckst) [reply docs/index.rst:163]: In MDA we have `has_positions`, `has_velocities`, `has_forces` flags on `Timesteps`. Would it make sense to have a bit array for all content in TNG trajectory that, for each queried step, just has a True or False for content present?

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:202]: sorry for delay, will get to this asap.

**Comment by orbeckst** [comment]
@hmacdope please ping me when you need me to look at things — this seems so far along that it's already in a superposition of merged and umerged ("Schrödinger's PR").

File: docs/index.rst
Comment (hmacdope) [reply docs/index.rst:163]: Makes sense and could be part of enhancements surrounding iteration. I'll raise an issue.

Comment (hmacdope) [reply docs/index.rst:163]: see #47

**Comment by hmacdope** [comment]
Hey @orbeckst, ready for re-review (I think).
I have added in some more explanation about calls to `read_success`.

I think we should possibly press ahead with this PR and add new features iterativley? I have raised a few new issues that we can address.

File: docs/index.rst
Comment (orbeckst) [inline docs/index.rst:208]: long line and "exception", "indicates" (spell check ;-) )

**Review by orbeckst — APPROVED** [review]
I am ok with merging this PR and forging ahead.

I found two typos but that's not a blocking request... I am not a big fan of having ``False`` for "no data" and "file corruption", "end of file" etc but for right now it's consistent with the underlying library and we could improve later.

Thanks for raising the follow-up issue.

**Comment by orbeckst** [comment]
@hmacdope are you able to merge the PR yourself once you feel it's ready?

**Comment by hmacdope** [comment]
@orbeckst sure am. I'll fix those typos.

**Pull request merged**
